File: android-app/feature-blockchain/src/main/java/com/chainlesschain/android/feature/blockchain/domain/model/Wallet.kt
package com.chainlesschain.android.feature.blockchain.domain.model

import com.chainlesschain.android.core.blockchain.model.SupportedChain
import kotlinx.serialization.Serializable

/**
 * Wallet domain model
 */
@Serializable
data class Wallet(
    val id: String,
    val name: String,
    val address: String,
    val chain: SupportedChain,
    val walletType: WalletType,
    val provider: WalletProvider,
    val derivationPath: String? = null,
    val isDefault: Boolean = false,
    val createdAt: Long = System.currentTimeMillis(),
    val updatedAt: Long = System.currentTimeMillis(),
    val metadata: WalletMetadata? = null
) {
    /**
     * Get shortened address for display
     */
    fun shortAddress(prefixLength: Int = 6, suffixLength: Int = 4): String {
        return if (address.length > prefixLength + suffixLength + 3) {
            "${address.take(prefixLength)}...${address.takeLast(suffixLength)}"
        } else {
            address
        }
    }

    /**
     * Check if wallet is internal (managed by app)
     */
    val isInternal: Boolean
        get() = walletType == WalletType.INTERNAL

    /**
     * Check if wallet is external (e.g., WalletConnect)
     */
    val isExternal: Boolean
        get() = walletType == WalletType.EXTERNAL
}

/**
 * Wallet type
 */
@Serializable
enum class WalletType {
    INTERNAL,   // Created/imported within the app
    EXTERNAL    // Connected via WalletConnect, etc.
}

/**
 * Wallet provider
 */
@Serializable
enum class WalletProvider {
    BUILTIN,        // App's built-in wallet
    WALLET_CONNECT, // WalletConnect
    METAMASK,       // MetaMask (via WalletConnect)
    TRUST_WALLET,   // Trust Wallet
    COINBASE,       // Coinbase Wallet
    LEDGER,         // Ledger hardware wallet
    OTHER
}

/**
 * Wallet metadata
 */
@Serializable
data class WalletMetadata(
    val iconUrl: String? = null,
    val accountIndex: Int = 0,
    val addressIndex: Int = 0,
    val isHardwareWallet: Boolean = false,
    val lastSyncedAt: Long? = null,
    val tags: List<String> = emptyList()
)

/**
 * Wallet balance
 */
@Serializable
data class WalletBalance(
    val walletId: String,
    val chain: SupportedChain,
    val nativeBalance: String,          // Raw balance in wei/smallest unit
    val nativeBalanceFormatted: String, // Formatted balance (e.g., "1.5 ETH")
    val usdValue: Double? = null,
    val tokens: List<TokenBalance> = emptyList(),
    val updatedAt: Long = System.currentTimeMillis()
)

/**
 * Token balance within a wallet
 */
@Serializable
data class TokenBalance(
    val token: Token,
    val balance: String,            // Raw balance
    val balanceFormatted: String,   // Formatted balance
    val usdValue: Double? = null
)

/**
 * Wallet with balance info
 */
data class WalletWithBalance(
    val wallet: Wallet,
    val balance: WalletBalance?
)

/**
 * Wallet creation options
 */
@Serializable
data class CreateWalletOptions(
    val name: String,
    val chain: SupportedChain = SupportedChain.ETHEREUM,
    val mnemonicStrength: Int = 12, // 12 or 24 words
    val passphrase: String = "",
    val accountIndex: Int = 0,
    val setAsDefault: Boolean = false
)

/**
 * Wallet import options
 */
@Serializable
data class ImportWalletOptions(
    val name: String,
    val chain: SupportedChain = SupportedChain.ETHEREUM,
    val importType: ImportType,
    val value: String, // Mnemonic, private key, or keystore JSON
    val password: String? = null, // For keystore
    val passphrase: String = "", // BIP39 passphrase
    val derivationPath: String? = null,
    val setAsDefault: Boolean = false
)

/**
 * Import type
 */
@Serializable
enum class ImportType {
    MNEMONIC,
    PRIVATE_KEY,
    KEYSTORE
}

File: ios-app/ChainlessChain/Features/Blockchain/Models/Wallet.swift
import Foundation

/// 钱包类型
enum WalletType: String, Codable {
    case internal       // 内置钱包
    case external       // 外部钱包
}

/// 钱包提供者
enum WalletProvider: String, Codable {
    case builtin        // 内置
    case metamask       // MetaMask
    case walletConnect  // WalletConnect
}

/// 钱包模型
struct Wallet: Identifiable, Codable {
    let id: String
    let address: String
    let walletType: WalletType
    let provider: WalletProvider
    let derivationPath: String
    let chainId: Int
    var isDefault: Bool
    let createdAt: Date

    /// 格式化地址（简短显示）
    var formattedAddress: String {
        guard address.count > 10 else { return address }
        let start = address.prefix(6)
        let end = address.suffix(4)
        return "\(start)...\(end)"
    }

    /// 获取链配置
    var chain: SupportedChain? {
        SupportedChain(rawValue: chainId)
    }

    /// 获取网络配置
    var networkConfig: NetworkConfig? {
        guard let chain = chain else { return nil }
        return NetworkConfig.config(for: chain)
    }
}

/// 钱包余额
struct WalletBalance: Codable {
    let walletId: String
    let chainId: Int
    let balance: String      // Wei余额（字符串避免精度问题）
    let symbol: String
    let decimals: Int
    let updatedAt: Date

    /// 格式化余额（转换为ETH/MATIC等）
    var formattedBalance: String {
        guard let balanceDecimal = Decimal(string: balance) else { return "0.0" }
        let divisor = pow(Decimal(10), decimals)
        let result = balanceDecimal / divisor
        return String(describing: result)
    }

    /// 显示余额（最多6位小数）
    var displayBalance: String {
        guard let value = Double(formattedBalance) else { return "0.0" }
        return String(format: "%.6f", value)
    }
}

/// 加密的钱包数据（存储在Keychain中）
struct EncryptedWalletData: Codable {
    let walletId: String
    let encryptedPrivateKey: String
    let encryptedMnemonic: String?
    let salt: String
    let iv: String
}

/// 钱包创建结果
struct WalletCreationResult {
    let wallet: Wallet
    let mnemonic: String        // 明文助记词（供用户备份）
    let privateKey: String      // 明文私钥（临时使用）
}

/// 钱包导入类型
enum WalletImportType {
    case mnemonic(String)       // 助记词
    case privateKey(String)     // 私钥
}
